File: run_test.cpp
#include "opencv2/opencv.hpp"
#include <iostream>
#include <chrono>

using namespace std;
using namespace cv;
using namespace std::chrono;

int main(){
    
    cout << "Loading video" << endl;
    // Create a VideoCapture object and open the input file
    // If the input is the web camera, pass 0 instead of the video file name
    VideoCapture video("./tracking_test0.avi");  //Name must include a number
    auto current_time = steady_clock::now();

    // Check if camera opened successfully
    if(!video.isOpened()){
        cout << "Error opening video stream or file" << endl;
        return -1;
        }
    else{
        cout << "Video open" << endl;
        }

    // Load pretrained XML classifier
    CascadeClassifier detector; 
    double scale=1;
    detector.load("haarcascade_frontalface_default.xml") ; 

    // Main loop
    while(true){
 
        Mat frame;  // OpenCV Matrix
        video >> frame; // Retrieve frame-by-frame
      
        // If the frame is empty, break
        if (frame.empty())
          break;
        
        Mat frame_gray;
        cvtColor( frame, frame_gray, COLOR_BGR2GRAY );

        // Perform detection
        vector<Rect> faces;
        detector.detectMultiScale( frame_gray, faces,1.1,10);
        for(size_t i = 0; i < faces.size(); i++) {
            rectangle(frame, faces[i], Scalar(0, 255, 0), 3);
            }

        // Record performance
        int perform_fps =  1/(duration<double, std::milli>(steady_clock::now()-current_time).count()/1e+3);
        current_time = steady_clock::now();
        putText(frame, //target image
                to_string(perform_fps)+" FPS", //text
                cv::Point(frame.cols/10*7, frame.rows/10*9), //top-left position
                cv::FONT_HERSHEY_DUPLEX,
                1,
                CV_RGB(0, 0, 0), //font color
                2);
    
        // Display the resulting frame
        imshow("Frame", frame );
    
        // Press  ESC on keyboard to exit
        char c=(char)waitKey(1);
        if(c==27){
            break;
            }
        } 
      
        video.release(); // When everything done, release the video capture object
        destroyAllWindows(); // Closes all the frames
      
        return 0;
        }

File: run_test.py
#!/usr/bin/env python3
"""Start object detector test"""

import cv2
import time
import sys

print("Welcome to the Object Detector Tester")
run_time = time.perf_counter()

print("Initialising detector")
faceDetector = cv2.CascadeClassifier(cv2.data.haarcascades + 'haarcascade_frontalface_default.xml')

print("Load video stream")
cap = cv2.VideoCapture(sys.argv[1])
font = cv2.FONT_HERSHEY_DUPLEX
color = (0, 0, 0) 
 
# Read until video is completed
while(cap.isOpened()):

  # Capture frame-by-frame
  ret, frame = cap.read()

  if ret == True:
    frameGray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    scale_factor = 1.1  # how much the image size is reduced at each image scale
    min_neighbours = 10 # how many neighbors each candidate rectangle should have to retain it
    detections = faceDetector.detectMultiScale(frameGray, scale_factor, min_neighbours)

    # Record performance
    rate_fps = 1/(time.perf_counter()-run_time)
    run_time = time.perf_counter()
    frame = cv2.putText(frame, str(round(rate_fps))+" FPS", ((frame.shape[1]//10)*7,(frame.shape[0]//10)*9), font, 1, color)

    # Add bounding boxes
    for (x, y, w, h) in detections:
        frame = cv2.rectangle(frame, (x, y), (x + w, y + h), (0, 255, 0), 3)

    # Display the resulting frame
    cv2.imshow('Frame',frame)
 
    # Press Q on keyboard to  exit
    if cv2.waitKey(1) & 0xFF == ord('q'):
      break
 
  # Break the loop
  else: 
    break

cap.release() # When everything done, release the video capture object
cv2.destroyAllWindows() # Closes all the frames

print("Complete")
